project/VersionUtils.scala

```diff
@@ -0,0 +1,75 @@
+import Mima.getMajorMinorPatch
+
+object VersionUtils {
+
+  ////////////////////////
+  // Java version utils //
+  ////////////////////////
+
+  val DEFAULT_JAVA_VERSION = 8
+  val JAVA_VERSION_FOR_SPARK_MASTER = 17
+  val JAVA_VERSION_FOR_LATEST_RELEASED_SPARK = DEFAULT_JAVA_VERSION
+
+  /** Java version of the current build. */
+  lazy val buildJavaVersion: Int = {
+    val versionStr = sys.props.getOrElse("java.version", "Unknown")
+    if (versionStr.startsWith("1.")) {
+      // Java 8 or lower: 1.6.0_23, 1.7.0, 1.7.0_80, 1.8.0_211
+      versionStr.substring(2, 3).toInt
+    } else {
+      // Java 9 or higher: 9.0.1, 11.0.4, 12, 12.0.1
+      val dot = versionStr.indexOf(".")
+      if (dot != -1) versionStr.substring(0, dot).toInt
+      else throw new IllegalArgumentException(s"Cannot parse Java version: $versionStr")
+    }
+  }
+
+  /**
+   * Get the Java compiler options that based on the target Java version, while
+   * also validating that the buld Java version is compatible with the target.
+   */
+  def getJavacOptionForTargetJavaVersion(targetJavaVersion: Int): Seq[String] = {
+    assert(buildJavaVersion >= targetJavaVersion,
```

Comment:
this the crux of the PR. 
this will validate and throw a clear message when any module is being built with insufficient java version.

Reply:
and it will also print exactly what is being used. its 1 line per module .. so its i think this verbosity is worth it when have a complex set up like ours.